cpp/include/raft/core/resource/device_memory_resource.hpp

```diff
@@ -41,36 +66,161 @@ class device_memory_resource : public resource {
  */
 class workspace_resource_factory : public resource_factory {
  public:
-  workspace_resource_factory(rmm::mr::device_memory_resource* mr_ = nullptr) : mr(mr_) {}
-  resource_type get_resource_type() override { return resource_type::WORKSPACE_RESOURCE; }
-  resource* make_resource() override { return new device_memory_resource(mr); }
+  explicit workspace_resource_factory(
+    std::shared_ptr<rmm::mr::device_memory_resource> mr = {nullptr},
+    std::optional<std::size_t> allocation_limit         = std::nullopt,
+    std::optional<std::size_t> alignment                = std::nullopt)
+    : allocation_limit_(allocation_limit.value_or(default_allocation_limit())),
+      alignment_(alignment),
+      mr_(mr ? mr : default_plain_resource())
+  {
+  }
+
+  auto get_resource_type() -> resource_type override { return resource_type::WORKSPACE_RESOURCE; }
+  auto make_resource() -> resource* override
+  {
+    return new limiting_memory_resource(mr_, allocation_limit_, alignment_);
+  }
+
+  /** Construct a sensible default pool memory resource. */
+  static inline auto default_pool_resource(std::size_t limit)
+    -> std::shared_ptr<rmm::mr::device_memory_resource>
+  {
+    // Set the default granularity to 1 GiB
+    constexpr std::size_t kOneGb = 1024lu * 1024lu * 1024lu;
+    // The initial size of the pool. The choice of this value only affects the performance a little
+    // bit. Heuristics:
+    //   1) the pool shouldn't be too big from the beginning independently of the limit;
+    //   2) otherwise, set it to half the max size to avoid too many resize calls.
+    auto min_size = std::min<std::size_t>(kOneGb, limit / 2lu);
+    // The pool is going to be place behind the limiting resource adaptor. This means the user won't
+    // be able to allocate more than 'limit' bytes of memory anyway. At the same time, the pool
+    // itself may consume a little bit more memory than the 'limit' due to memory fragmentation.
+    // Therefore, we look for a compromise, such that:
+    //   1) 'limit' is accurate - the user should be more likely to run into the limiting
+    //      resource adaptor bad_alloc error than into the pool bad_alloc error.
+    //   2) The pool doesn't grab too much memory on top of the 'limit'.
+    auto max_size = std::min<std::size_t>(limit + kOneGb / 2lu, limit * 3lu / 2lu);
+    auto upstream = rmm::mr::get_current_device_resource();
+    return std::make_shared<rmm::mr::pool_memory_resource<rmm::mr::device_memory_resource>>(
+      upstream, min_size, max_size);
+  }
+
+  /** Get the global memory resource wrapped into an unmanaged shared_ptr (with no deleter). */
```

Comment:
Please add a note why no deleter is used (we have discussed in [this thread](https://github.com/rapidsai/raft/pull/1356#discussion_r1150360407), but it is better to record this information in the code comments).